Polyfills/Window/Source/Window.cpp

```diff
@@ -72,17 +72,50 @@ namespace Babylon::Polyfills::Internal
     Window::Window(const Napi::CallbackInfo& info)
         : Napi::ObjectWrap<Window>{info}
         , m_runtime{JsRuntime::GetFromJavaScript(info.Env())}
+        , m_timeoutDispatcher{m_runtime}
     {
     }
 
-    void Window::SetTimeout(const Napi::CallbackInfo& info)
+    Napi::Value Window::SetTimeout(const Napi::CallbackInfo& info)
     {
-        auto function = Napi::Persistent(info[0].As<Napi::Function>());
-        auto milliseconds = std::chrono::milliseconds{info[1].As<Napi::Number>().Int32Value()};
-
         auto& window = *static_cast<Window*>(info.Data());
 
-        window.RecursiveWaitOrCall(std::make_shared<Napi::FunctionReference>(std::move(function)), std::chrono::system_clock::now() + milliseconds);
+        const auto arg0 = info[0];
+        if (!arg0.IsFunction())
+        {
+            return Napi::Value::From(info.Env(), window.m_timeoutDispatcher->NextTimeoutId());
+        }
+        auto function = arg0.As<Napi::Function>();
+
+        std::chrono::milliseconds delay{0};
+        if (info[1].IsString())
+        {
+            const auto delayString = info[1].As<Napi::String>().Utf8Value();
+            try
+            {
+                delay = std::chrono::milliseconds{std::stoi(delayString)};
+            }
+            catch (std::invalid_argument&)
+            {
+            }
+        }
+        else if (info[1].IsNumber())
+        {
+            delay = std::chrono::milliseconds{info[1].ToNumber().Int32Value()};
+        }
```

Comment:
N-API has a method called `ToNumber()` which coerces the type to be a number. Does that work?

Reply:
Ya, I was using `ToNumber()` earlier but it wasn't working with the JSI engine.

Reply:
ToNumber isn't completely implemented in napi-jsi, but we can probably make it work.

Reply:
https://github.com/BabylonJS/BabylonNative/pull/1146/commits/e83cca81c7de01aa051e2d3b453b637ba53a3698